File: sdk/runanywhere-flutter/lib/capabilities/memory/cache_eviction.dart
import '../../foundation/logging/sdk_logger.dart';
import 'allocation_manager.dart';
import 'models/memory_models.dart';

/// Manages cache eviction strategies and model selection for memory cleanup
/// Matches iOS CacheEviction
class CacheEviction {
  final SDKLogger _logger = SDKLogger(category: 'CacheEviction');
  AllocationManager? _allocationManager;

  void setAllocationManager(AllocationManager manager) {
    _allocationManager = manager;
  }

  // MARK: - Model Selection for Eviction

  /// Select models to evict to meet target memory
  List<String> selectModelsToEvict({required int targetMemory}) {
    final models = _getCurrentModels();
    return _selectModelsUsingStrategy(
      models: models,
      targetMemory: targetMemory,
      aggressive: false,
    );
  }

  /// Select models for critical eviction (more aggressive)
  List<String> selectModelsForCriticalEviction({required int targetMemory}) {
    final models = _getCurrentModels();
    return _selectModelsUsingStrategy(
      models: models,
      targetMemory: targetMemory,
      aggressive: true,
    );
  }

  /// Select a specific count of models to evict
  List<String> selectModelsToEvictByCount({required int count}) {
    final models = _getCurrentModels();
    final sortedModels =
        _sortModelsByEvictionPriority(models, aggressive: false);
    return sortedModels.take(count).map((m) => m.model.id).toList();
  }

  /// Select least important models
  List<String> selectLeastImportantModels({required int maxCount}) {
    final models = _getCurrentModels();
    final sortedModels = _sortModelsByImportance(models);
    return sortedModels.take(maxCount).map((m) => m.model.id).toList();
  }

  // MARK: - Eviction Strategies

  List<String> _selectModelsUsingStrategy({
    required List<MemoryLoadedModelInfo> models,
    required int targetMemory,
    required bool aggressive,
  }) {
    // Default to least recently used strategy
    return _selectByLeastRecentlyUsed(
      models: models,
      targetMemory: targetMemory,
      aggressive: aggressive,
    );
  }

  List<String> _selectByLeastRecentlyUsed({
    required List<MemoryLoadedModelInfo> models,
    required int targetMemory,
    required bool aggressive,
  }) {
    final sortedModels = models.toList()
      ..sort((a, b) => a.lastUsed.compareTo(b.lastUsed));
    return _selectModelsToTarget(
      sortedModels: sortedModels,
      targetMemory: targetMemory,
      aggressive: aggressive,
    );
  }

  // ignore: unused_element - Reserved for future use (matches iOS eviction strategies)
  List<String> _selectByLargestFirst({
    required List<MemoryLoadedModelInfo> models,
    required int targetMemory,
    required bool aggressive,
  }) {
    final sortedModels = models.toList()
      ..sort((a, b) => b.size.compareTo(a.size));
    return _selectModelsToTarget(
      sortedModels: sortedModels,
      targetMemory: targetMemory,
      aggressive: aggressive,
    );
  }

  // ignore: unused_element - Reserved for future use (matches iOS eviction strategies)
  List<String> _selectByOldestFirst({
    required List<MemoryLoadedModelInfo> models,
    required int targetMemory,
    required bool aggressive,
  }) {
    final sortedModels = models.toList()
      ..sort((a, b) => a.lastUsed.compareTo(b.lastUsed));
    return _selectModelsToTarget(
      sortedModels: sortedModels,
      targetMemory: targetMemory,
      aggressive: aggressive,
    );
  }

  // ignore: unused_element - Reserved for future use (matches iOS eviction strategies)
  List<String> _selectByPriority({
    required List<MemoryLoadedModelInfo> models,
    required int targetMemory,
    required bool aggressive,
  }) {
    final sortedModels = models.toList()
      ..sort((a, b) {
        // Lower priority models are evicted first
        if (a.priority != b.priority) {
          return a.priority.value.compareTo(b.priority.value);
        }
        // If same priority, evict least recently used first
        return a.lastUsed.compareTo(b.lastUsed);
      });
    return _selectModelsToTarget(
      sortedModels: sortedModels,
      targetMemory: targetMemory,
      aggressive: aggressive,
    );
  }

  List<String> _selectModelsToTarget({
    required List<MemoryLoadedModelInfo> sortedModels,
    required int targetMemory,
    required bool aggressive,
  }) {
    final modelsToEvict = <String>[];
    int freedMemory = 0;

    for (final model in sortedModels) {
      // In non-aggressive mode, skip critical priority models unless absolutely necessary
      if (!aggressive &&
          model.priority == MemoryPriority.critical &&
          freedMemory > 0) {
        continue;
      }

      modelsToEvict.add(model.model.id);
      freedMemory += model.size;

      final sizeString = _formatBytes(model.size);
      _logger.debug(
          "Selected model '${model.model.name}' for eviction (size: $sizeString)");

      if (freedMemory >= targetMemory) {
        break;
      }
    }

    final targetMemoryString = _formatBytes(targetMemory);
    _logger.info(
        'Selected ${modelsToEvict.length} models for eviction, target memory: $targetMemoryString');

    return modelsToEvict;
  }

  // MARK: - Model Sorting

  List<MemoryLoadedModelInfo> _sortModelsByEvictionPriority(
    List<MemoryLoadedModelInfo> models, {
    required bool aggressive,
  }) {
    return models.toList()
      ..sort((a, b) {
        // In aggressive mode, ignore critical priority
        if (!aggressive) {
          if (a.priority != b.priority) {
            return a.priority.value.compareTo(b.priority.value);
          }
        }

        // Consider both recency and size
        final aScore = _calculateEvictionScore(a);
        final bScore = _calculateEvictionScore(b);

        return aScore
            .compareTo(bScore); // Lower score = higher eviction priority
      });
  }

  List<MemoryLoadedModelInfo> _sortModelsByImportance(
      List<MemoryLoadedModelInfo> models) {
    return models.toList()
      ..sort((a, b) {
        // Higher priority = more important (lower eviction priority)
        if (a.priority != b.priority) {
          return a.priority.value.compareTo(b.priority.value);
        }

        // More recently used = more important
        return a.lastUsed.compareTo(b.lastUsed);
      });
  }

  double _calculateEvictionScore(MemoryLoadedModelInfo model) {
    final timeSinceUse = DateTime.now().difference(model.lastUsed).inSeconds;
    final priorityWeight =
        model.priority.value * 1000; // Higher priority = higher score
    final recencyScore = timeSinceUse / 3600; // Hours since last use

    // Lower score = higher eviction priority
    return priorityWeight - recencyScore;
  }

  // MARK: - Model Information

  /// Get eviction candidates with minimum memory size
  List<MemoryLoadedModelInfo> getEvictionCandidates({required int minMemory}) {
    final models = _getCurrentModels();
    return models.where((m) => m.size >= minMemory).toList();
  }

  /// Get models by priority
  List<MemoryLoadedModelInfo> getModelsByPriority(MemoryPriority priority) {
    final models = _getCurrentModels();
    return models.where((m) => m.priority == priority).toList();
  }

  /// Get models older than interval
  List<MemoryLoadedModelInfo> getModelsByUsageAge(
      {required Duration olderThan}) {
    final models = _getCurrentModels();
    final cutoffDate = DateTime.now().subtract(olderThan);
    return models.where((m) => m.lastUsed.isBefore(cutoffDate)).toList();
  }

  // MARK: - Statistics

  /// Get eviction statistics
  EvictionStatistics getEvictionStatistics() {
    final models = _getCurrentModels();

    final totalMemory = models.fold<int>(0, (sum, m) => sum + m.size);
    final modelsByPriority = <MemoryPriority, int>{};
    for (final model in models) {
      modelsByPriority[model.priority] =
          (modelsByPriority[model.priority] ?? 0) + 1;
    }

    final avgLastUsed = models.isEmpty
        ? DateTime.now()
        : DateTime.fromMillisecondsSinceEpoch(
            (models
                        .map((m) => m.lastUsed.millisecondsSinceEpoch)
                        .reduce((a, b) => a + b) /
                    models.length)
                .round(),
          );

    final oldestModel = models.isEmpty
        ? DateTime.now()
        : models
            .reduce((a, b) => a.lastUsed.isBefore(b.lastUsed) ? a : b)
            .lastUsed;

    final largestModel = models.isEmpty
        ? 0
        : models.reduce((a, b) => a.size > b.size ? a : b).size;

    return EvictionStatistics(
      totalModels: models.length,
      totalMemory: totalMemory,
      modelsByPriority: modelsByPriority,
      averageLastUsed: avgLastUsed,
      oldestModel: oldestModel,
      largestModel: largestModel,
    );
  }

  // MARK: - Private Implementation

  List<MemoryLoadedModelInfo> _getCurrentModels() {
    // Get models from allocation manager
    return _allocationManager?.getLoadedModelInfos() ?? [];
  }

  String _formatBytes(int bytes) {
    if (bytes < 1024) return '$bytes B';
    if (bytes < 1024 * 1024) return '${(bytes / 1024).toStringAsFixed(1)} KB';
    if (bytes < 1024 * 1024 * 1024) {
      return '${(bytes / (1024 * 1024)).toStringAsFixed(1)} MB';
    }
    return '${(bytes / (1024 * 1024 * 1024)).toStringAsFixed(1)} GB';
  }
}

File: sdk/runanywhere-react-native/src/Capabilities/Memory/Services/CacheEviction.ts
/**
 * CacheEviction.ts
 *
 * Manages cache eviction strategies and model selection for memory cleanup
 *
 * Reference: sdk/runanywhere-swift/Sources/RunAnywhere/Capabilities/Memory/Services/CacheEviction.swift
 */

import type {
  MemoryLoadedModelInfo,
  MemoryPriority,
  EvictionStatistics,
} from '../../../Core/Protocols/Memory/MemoryModels';
import type { AllocationManager } from './AllocationManager';

/**
 * Manages cache eviction strategies and model selection for memory cleanup
 */
export class CacheEviction {
  private allocationManager: AllocationManager | null = null;

  /**
   * Set allocation manager
   */
  public setAllocationManager(manager: AllocationManager): void {
    this.allocationManager = manager;
  }

  /**
   * Select models to evict
   */
  public selectModelsToEvict(targetMemory: number): string[] {
    const models = this.getCurrentModels();
    return this.selectModelsUsingStrategy(
      models,
      targetMemory,
      false
    );
  }

  /**
   * Select models for critical eviction
   */
  public selectModelsForCriticalEviction(targetMemory: number): string[] {
    const models = this.getCurrentModels();
    return this.selectModelsUsingStrategy(models, targetMemory, true);
  }

  /**
   * Select models to evict by count
   */
  public selectModelsToEvict(count: number): string[] {
    const models = this.getCurrentModels();
    const sortedModels = this.sortModelsByEvictionPriority(models, false);
    return sortedModels.slice(0, count).map((m) => m.model.id);
  }

  /**
   * Select least important models
   */
  public selectLeastImportantModels(maxCount: number): string[] {
    const models = this.getCurrentModels();
    const sortedModels = this.sortModelsByImportance(models);
    return sortedModels.slice(0, maxCount).map((m) => m.model.id);
  }

  /**
   * Select models using strategy
   */
  private selectModelsUsingStrategy(
    models: MemoryLoadedModelInfo[],
    targetMemory: number,
    aggressive: boolean
  ): string[] {
    // Default to least recently used strategy
    return this.selectByLeastRecentlyUsed(
      models,
      targetMemory,
      aggressive
    );
  }

  /**
   * Select by least recently used
   */
  private selectByLeastRecentlyUsed(
    models: MemoryLoadedModelInfo[],
    targetMemory: number,
    aggressive: boolean
  ): string[] {
    const sortedModels = models.sort(
      (a, b) => a.lastUsed.getTime() - b.lastUsed.getTime()
    );
    return this.selectModelsToTarget(sortedModels, targetMemory, aggressive);
  }

  /**
   * Select models to target
   */
  private selectModelsToTarget(
    sortedModels: MemoryLoadedModelInfo[],
    targetMemory: number,
    aggressive: boolean
  ): string[] {
    const modelsToEvict: string[] = [];
    let freedMemory = 0;

    for (const model of sortedModels) {
      // In non-aggressive mode, skip critical priority models unless absolutely necessary
      if (
        !aggressive &&
        model.priority === MemoryPriority.Critical &&
        freedMemory > 0
      ) {
        continue;
      }

      modelsToEvict.push(model.model.id);
      freedMemory += model.size;

      if (freedMemory >= targetMemory) {
        break;
      }
    }

    return modelsToEvict;
  }

  /**
   * Sort models by eviction priority
   */
  private sortModelsByEvictionPriority(
    models: MemoryLoadedModelInfo[],
    aggressive: boolean
  ): MemoryLoadedModelInfo[] {
    return models.sort((lhs, rhs) => {
      // In aggressive mode, ignore critical priority
      if (!aggressive) {
        if (lhs.priority !== rhs.priority) {
          return lhs.priority - rhs.priority;
        }
      }

      // Consider both recency and size
      const lhsScore = this.calculateEvictionScore(lhs);
      const rhsScore = this.calculateEvictionScore(rhs);

      return lhsScore - rhsScore; // Lower score = higher eviction priority
    });
  }

  /**
   * Sort models by importance
   */
  private sortModelsByImportance(
    models: MemoryLoadedModelInfo[]
  ): MemoryLoadedModelInfo[] {
    return models.sort((lhs, rhs) => {
      // Higher priority = more important (lower eviction priority)
      if (lhs.priority !== rhs.priority) {
        return lhs.priority - rhs.priority;
      }

      // More recently used = more important
      return lhs.lastUsed.getTime() - rhs.lastUsed.getTime();
    });
  }

  /**
   * Calculate eviction score
   */
  private calculateEvictionScore(model: MemoryLoadedModelInfo): number {
    const timeSinceUse =
      (Date.now() - model.lastUsed.getTime()) / 1000; // seconds
    const priorityWeight = model.priority * 1000; // Higher priority = higher score
    const recencyScore = timeSinceUse / 3600; // Hours since last use

    // Lower score = higher eviction priority
    return priorityWeight - recencyScore;
  }

  /**
   * Get eviction candidates
   */
  public getEvictionCandidates(minMemory: number): MemoryLoadedModelInfo[] {
    const models = this.getCurrentModels();
    return models.filter((m) => m.size >= minMemory);
  }

  /**
   * Get models by priority
   */
  public getModelsByPriority(
    priority: MemoryPriority
  ): MemoryLoadedModelInfo[] {
    const models = this.getCurrentModels();
    return models.filter((m) => m.priority === priority);
  }

  /**
   * Get models by usage age
   */
  public getModelsByUsageAge(olderThan: number): MemoryLoadedModelInfo[] {
    const models = this.getCurrentModels();
    const cutoffDate = new Date(Date.now() - olderThan * 1000);
    return models.filter((m) => m.lastUsed < cutoffDate);
  }

  /**
   * Get eviction statistics
   */
  public getEvictionStatistics(): EvictionStatistics {
    const models = this.getCurrentModels();

    const totalMemory = models.reduce((sum, m) => sum + m.size, 0);
    const modelsByPriority: { [key: number]: number } = {};
    for (const model of models) {
      modelsByPriority[model.priority] =
        (modelsByPriority[model.priority] || 0) + 1;
    }

    const avgLastUsed =
      models.length > 0
        ? new Date(
            models.reduce(
              (sum, m) => sum + m.lastUsed.getTime(),
              0
            ) / models.length
          )
        : new Date();

    const oldestModel =
      models.length > 0
        ? new Date(
            Math.min(...models.map((m) => m.lastUsed.getTime()))
          )
        : new Date();

    const largestModel =
      models.length > 0
        ? Math.max(...models.map((m) => m.size))
        : 0;

    return {
      totalModels: models.length,
      totalMemory,
      modelsByPriority,
      averageLastUsed: avgLastUsed,
      oldestModel,
      largestModel,
      totalMemoryString: this.formatBytes(totalMemory),
      largestModelString: this.formatBytes(largestModel),
    };
  }

  /**
   * Get current models
   */
  private getCurrentModels(): MemoryLoadedModelInfo[] {
    return this.allocationManager?.getLoadedModels() ?? [];
  }

  /**
   * Format bytes to human-readable string
   */
  private formatBytes(bytes: number): string {
    const units = ['B', 'KB', 'MB', 'GB', 'TB'];
    let size = bytes;
    let unitIndex = 0;

    while (size >= 1024 && unitIndex < units.length - 1) {
      size /= 1024;
      unitIndex++;
    }

    return `${size.toFixed(2)} ${units[unitIndex]}`;
  }
}

